### PR TITLE
Added "Useful tips" in learning section 

### PR DESCRIPTION
**Pull request by ste911**
The aim of this PR is adding some suggestions that can help a user in the process of creation or modification of an operator.
Those are all practical suggestions related to coding itself.
They have been added in learning/operator/ section of the site.

### CODE REVIEW

File: content/learning/operator/operator.md [hunk 1]
```diff
@@ -69,6 +69,54 @@ The kubebuilder framework uses the controller-runtime library to communicate wit
...
+func (r *_ObjectToReconcile_) SetupWithManager(mgr ctrl.Manager) error {
+	
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&principal_resource).
+		Owns(&second resource_to_watch).
+		Complete(r)
+}
+```
+
+2. It is possible to filter the case of reconcile that our operator perform thanks to the [Predicates](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/). To use them you can modify the setup function explayned above:
```
Comment (giorio94) [inline content/learning/operator/operator.md:86]: * Please, give an example why you would do this and explain the example code. Remember to put yourself in the shoes of a newbie reader that knows nothing about how to build an operator, not to mention the advanced aspects. Would you understand? Keep iterating and improving the documentation until the answer is yes.
* *filter the case of reconcile* -> what do you mean? Absolutely not clear to me
* explayned -> explained

File: content/learning/operator/operator.md [hunk 2]
```diff
@@ -69,6 +69,54 @@ The kubebuilder framework uses the controller-runtime library to communicate wit
...
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+      		// Here you can compare old version and new version of the watched object
+			return condition_to_evaluate
+		},
+	}
+}
+```
+
+3. Another problem that you can face is about the  [CreateOrUpdate](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate) function in which you have to create a initial resource (composed only by ObjectMeta's data) before the function's call and, inside it, with different functions, you call a mutation functions in which it is possible to modify the resource fields; this function, is intrinsically idempotent.
```
Comment (giorio94) [inline content/learning/operator/operator.md:114]: Same as above. Start giving the context of what you want to do, then explain how and finally give an example.

File: content/learning/operator/operator.md [hunk 3]
```diff
@@ -69,6 +69,54 @@ The kubebuilder framework uses the controller-runtime library to communicate wit
...
+
+
+4. You may have some little problem using update functions of controller-utils package (CreateOrUpdate(), Update(), Status().Update()).
```
Comment (giorio94) [inline content/learning/operator/operator.md:117]: Again, the same. What does *some little problem* means? How could a reader know whether this is the problem he is experiencing?

File: content/learning/operator/operator.md [hunk 4]
```diff
@@ -69,6 +69,54 @@ The kubebuilder framework uses the controller-runtime library to communicate wit
...
+This is due to the fact that all those function control the version number of the resources to be updated. To overcome this problem it is possibile to use "Patch Version" of those: CreateOrPatch(), Patch(), Status().Patch() that do not make any check on the version.
```
Comment (giorio94) [inline content/learning/operator/operator.md:118]: Is it good that the version is not checked? What issues could it lead to? Please add also an example of how a patch operation can be performed (it does not seem intuitive to me).

**Review by giorio94 — COMMENTED** [review]
Some comments inline

File: content/learning/operator/operator.md [hunk 5]
```diff
@@ -69,6 +69,54 @@ The kubebuilder framework uses the controller-runtime library to communicate wit
 4. the [request](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Request "request") received in the reconciler function
 5. k8s [corev1](https://godoc.org/k8s.io/api/core/v1 "corev1") in case you need to interact with core k8s kind (pods, deployments, configmaps etc)
 
+## Useful tips
+
+Here are some useful suggestion if you are new to programming an operator:
+1. You may need to watch a second resource. In this case what you need to do is to add Owns in the [setup](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#setup) function :
```
[the remark this answers is not in the document]
Comment (ste911) [reply content/learning/operator/operator.md]: Thank you for all you suggestions, we have re-organized the all documentation

File: content/learning/operator/operator.md [hunk 6]
```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
 4. the [request](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Request "request") received in the reconciler function
 5. k8s [corev1](https://godoc.org/k8s.io/api/core/v1 "corev1") in case you need to interact with core k8s kind (pods, deployments, configmaps etc)
 
+## Useful tips
+
+Here are some useful suggestions if you are new to programming an operator:
+
+1. You may need to watch more than one resource, e.g., when you have to react to different conditions.
+This leads the operator to wake up when either one of the watched resources is modified.
+A possible example is the following: your controller manages a CRD "laboratory" in which a label identifies the status (running or stopped) of a VirtualMachine (VM), which can be started or stopped directly by the user.
```
Comment (frisso) [inline content/learning/operator/operator.md:78]: In this example it seems that we're watching for a single resource. Can you clarify why this example applies to the case under consideration here, i.e., "You may need to watch more than one resource"?

File: content/learning/operator/operator.md [hunk 7]
```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+        return ctrl.SetControllerReference(resource, &vm, r.Scheme)
+	})
+```
+
+
+4. While updating a resources, sometimes applying an Update() on a resource could trigger errors like: "_operation cannot be fulfilled on resource, the object has been modified_".
+This means that the controller is using a different version of the resource from the one present in the cluster; this of course leads to a failure when updating the resource.
+To overcome this problem you can use the function Patch() instead of Update().
+In fact, Patch() does not make any check on the resource version; a possible problem consists in the possibility to lose some updates (since the changes in a resource are compared thanks to the version of the resource itself).
+However, the hypothetical loss of updates should not be a problem because of idempotence.
```
Comment (frisso) [inline content/learning/operator/operator.md:148]: This sentence is not clear to me. Why the idempotence prevents you from any losses?

File: content/learning/operator/operator.md (hunk 6 above)
Comment (ste911) [reply content/learning/operator/operator.md]: This example has been updated. Thank you for your suggestion!

File: content/learning/operator/operator.md (hunk 7 above)
Comment (ste911) [reply content/learning/operator/operator.md:148]: For sure the Patch function does not make any control to resource version. In theory a not aligned resource version can lead to an update loss. Unfortunately  we could not find any explanation of why it is possible to use that function without having troubles. What we thought is that since this problem happens only if you try to patch an older version of the cluster; the situation should be something like this: in the cluster you have the resource version 10 and you try to Patch it with a resource version 9. Thanks to the idempotence property of the operator it should not be a problem to update again the resource, without losing data at all.

File: content/learning/operator/operator.md [hunk 8]
```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+        },
+    }
+}
+```
+
+A Predicate takes an event of a particular type (Create / Update / Delete) and returns true if the event should be processed by reconcile.
+
+3. Another problem that you can face relates to the creation and the update of resources in the cluster.
+Given that a controller must be _idempotent_ (superfluous reconcile has no effect, e.g. resources already present in the cluster must be updated and not recreated every reconcile), it is better to use [CreateOrUpdate](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate), a function that is intrinsically idempotent and it is part of the `controller-utils` package.
+In order to use it you have to create an initial resource composed only by ObjectMeta's data; then you have to pass a _mutate function_ as a parameter, which represents a function that works on the resource itself for example adding labels, modifying annotations and describing specifications.
```
Comment (giorio94) [inline content/learning/operator/operator.md:131]:
```suggestion
To use the `CreateOrUpdate` function you have to first create an initial instance of the resource itself, specifying only its name and the namespace, if appropriate.
Then, the function takes care of initializing the content of the resource with the current version retrieved from the cluster (if any), before executing the  _mutate function_ specified as a parameter.
This has to configure the resource itself, e.g., adding labels, modifying annotations and describing specifications, to make it converge to the desired status.
Finally, `CreateOrUpdate` takes care of saving the modified resource in the cluster.
In other words, this function implements a very similar behavior to `kubectl apply`.
```

File: content/learning/operator/operator.md [hunk 9]
```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+
+Here there is a CreateOrUpdate example for a VirtualMachine:
+```
+vm := v1.VirtualMachine{ObjectMeta: metav1.ObjectMeta{Name: "name", Namespace: "namespace"}}
+    _, err = ctrl.CreateOrUpdate(ctx, r.Client, &vm, func() error {
+        UpdateVirtualMachineSpec(&vm)
+        UpdateLabels(vm.Spec.Template.ObjectMeta.Labels)
+        return ctrl.SetControllerReference(resource, &vm, r.Scheme)
+	})
```
Comment (giorio94) [inline content/learning/operator/operator.md:140]: First, check the indentation. Second, I would stick to an example involving core Kubernetes components (e.g. deployment, service, secret, configmap, ...), as VMs are custom resources which may not be well known and may raise confusion.

File: content/learning/operator/operator.md [hunk 10]
```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+```
+PatchedResource:= *resource.DeepCopy()
+PatchedResource.Status.IP = newIp
+PatchedResource.Status.Phase = newEventReason
+PatchedResource.Status.URL = newUrl
+
+if err := r.Status().Patch(ctx, &PatchedResource, client.MergeFrom(resource)); err != nil {
+	klog.Error("Unable to patch")
+	klog.Error(err)
+}
```
Comment (giorio94) [inline content/learning/operator/operator.md:159]: Same here about using core resources to make the workflow clearer.

File: content/learning/operator/operator.md (hunk 7 above)
Comment (giorio94) [reply content/learning/operator/operator.md:148]: I believe this actually depends on the context, as sometimes it may be appropriate to use a patch operation and in some others not, as it can lead to loss of data. I feel this mainly boils down to whether you are supposed to be the only owner of the resource (hence you may only overwrite your own changes, and it may not matter thanks to idempotence), or parallel changes are possible (hence you risk to loose data). At the same time, also the error mentioned at the beginning of this section may, or may not, be a problem, as the best solution may just be to get a new version of the resource and retry (e.g. in a subsequent reconcile). I believe these aspects should be better investigated and this section to be adapted accordingly.

**Review by giorio94 — COMMENTED** [review]
Some additional comments from my side.

Please, check also the appearance when rendered with Hugo, as there are some indentation problems with the numbered list.

File: content/learning/operator/operator.md (hunk 7 above)
Comment (ste911) [reply content/learning/operator/operator.md:148]: It would be very interesting to analyze these aspects. Unfortunately we have already searched in the Golang documentation and some forums, but there is very few materials about this (actually, we found nothing). For sure your analysis is much more better then our, but on the basis of our experience we really do not know how to improve it

File: content/learning/operator/operator.md [hunk 11]
```diff
@@ -65,10 +65,101 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+            },
+            DeleteFunc: func(e event.DeleteEvent) bool {
+                // Here you can compare old version and new version of the watched object
+                return condition_to_evaluate
+            },
+        }
+    }
+    ```
+
+    A Predicate takes an event of a particular type (Create / Update / Delete) and returns true if the event should be processed by reconcile. As you can see in this [link](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate "link"), predicates package already provides some predicates for common scenarios, such as to select only the modifications involving changes to the spec or filtering by label.
```
Comment (giorio94) [inline content/learning/operator/operator.md:124]:
```suggestion
    A Predicate takes an event of a particular type (Create / Update / Delete) and returns true if the event should be processed by reconcile. As you can see in this [link](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate "link"), the `predicates` package already provides some predicates for common scenarios, such as to select only the modifications involving changes to the spec or filtering by label.
```

File: content/learning/operator/operator.md [hunk 12]
```diff
@@ -65,10 +65,101 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+
+3. Another problem that you can face relates to the creation and the update of resources in the cluster.
+Given that a controller must be _idempotent_ (i.e., a superfluous reconcile should have no effect and additional resources should be created or updated depending on whether already present in the cluster)), it is better to use [CreateOrUpdate](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate), a function that is intrinsically idempotent and it is part of the `controller-utils` package.
```
Comment (giorio94) [inline content/learning/operator/operator.md:127]:
```suggestion
Given that a controller must be _idempotent_ (i.e., a superfluous reconcile should have no effect and additional resources should be created or updated depending on whether already present in the cluster), it is better to use [CreateOrUpdate](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate), a function that is intrinsically idempotent and it is part of the `controller-utils` package.
```

File: content/learning/operator/operator.md [hunk 13]
```diff
@@ -65,10 +65,101 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+            Name: "name", Namespace: "namespace"}
+            }
+    _, err = ctrl.CreateOrUpdate(ctx, r.Client, &deployment, func() error {
+            UpdateDeploymentSpec(&deployment)
+            return ctrl.SetControllerReference( &Owner, &deployment, r.Scheme)
+        })
+    ```
+
+
+4. While updating a resources, sometimes applying an Update() on a resource could trigger errors like: "_operation cannot be fulfilled on resource, the object has been modified_".
```
Comment (giorio94) [inline content/learning/operator/operator.md:147]:
```suggestion
4. While updating a resource, sometimes the `Update()` function could trigger errors like: "_operation cannot be fulfilled on resource, the object has been modified_".
```

File: content/learning/operator/operator.md [hunk 14]
```diff
@@ -65,10 +65,101 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
...
+This means that the controller is using a different version of the resource from the one present in the cluster; this of course leads to a failure when updating the resource.
+To overcome this problem you can use the function Patch() instead of Update().
+In fact, Patch() does not make any check on the resource version; a possible problem consists in the possibility to lose some updates (since the changes in a resource are compared thanks to the version of the resource itself).
+If the resource is watched only by one controller, you only may risk to overwrite your own changes and this may no matter thanks to idempotence. If the resource is owned by more than one controller, you should be very careful to use this function since you can loss some updates made by others controllers.
```
Comment (giorio94) [inline content/learning/operator/operator.md:151]:
```suggestion
If the resource is watched only by one controller, you only may risk to overwrite your own changes and this may not matter thanks to idempotence. If the resource is owned by more than one controller, you should be very careful to use this function since you can lose some updates made by others controllers.
```

**Review by giorio94 — APPROVED** [review]
/lgtm

Just a couple of very minor linguistic modifications.

**Comment by giorio94** [comment]
Please, squash the commits bifore merging

**Review by frisso — APPROVED** [review]

**Comment by frisso** [comment]
Thanks!
